Deferrable Operators Docs Edits (#33620)
mistercrunch/Airflow: docs/apache-airflow/authoring-and-scheduling/deferring.rst

Deferrable Operators & Triggers
===============================

> Standard :doc:`Operators </core-concepts/operators>` and :doc:`Sensors <../core-concepts/sensors>` take up a full *worker slot* for the entire time they are running, even if they are idle. For example, if you only have 100 worker slots available to run tasks, and you have 100 DAGs waiting on a sensor that's currently running but idle, then you *cannot run anything else* - even though your entire Airflow cluster is essentially idle. ``reschedule`` mode for sensors solves some of this, by allowing sensors to only run at fixed intervals, but it is inflexible and only allows using time as the reason to resume, not other criteria.

Reviewer: ```suggestion
Standard :doc:`Operators </core-concepts/operators>` and :doc:`Sensors <../core-concepts/sensors>` take up a full *worker slot* for the entire time they are running, even if they are idle. For example, if you only have 100 worker slots available to run tasks, and you have 100 DAGs waiting on a sensor that's currently running but idle, then you *cannot run anything else* - even though your entire Airflow cluster is essentially idle. ``reschedule`` mode for sensors solves some of this, by allowing sensors to only run at fixed intervals, but it is inflexible in that it only allows using time as the reason to resume, not other criteria.
```

> This is where *Deferrable Operators* can be used. A deferrable operator can suspend itself and free up the worker for other processes when certain conditions are met. A pre-defined *Trigger* then resumes the deferred operator. As a result, while it is suspended (deferred) and isn't using worker slot, your cluster has fewer resources wasted on idle operators or sensors. By default, deferred tasks don't use pool slots. If you would like them to, you can change this by editing the pool in question.

Reviewer 2: I don't really like this rewording: "for other processes when certain conditions are met. A pre-defined *Trigger* then resumes the deferred operator."

"when certain conditions are met" is technically correct, but is also a bit vague. Since all, or at least the vast majority only defer when they need to wait, maybe we can incorporate that so we have an example "condition"?

"A pre-defined *Trigger* then resumes the deferred operator" - I thought the old wording was easier to grok. The trigger is responsible for resuming the remaining non-async operator code (if any? @dstandish, did we ever get that short circuit done?).

Reviewer: overall i like where this is going, reducing word count and making it more economical and faster to digest.

@jedcunningham re "exit directly from trigger" we did not yet implement

Reviewer: you could probably chop "when certain conditions are met".  it's a bit ambiguous. in reality, a deferrable operator can defer whenever the heck it wants to.

Reviewer: but, typically, it defers because it has nothing to do but wait -- and that's why deferrable operators exist and that's what this is about.

Reviewer: here's a crack add trying to be more accurate, if not more clear
maybe it will help you in getting to the right balance
```suggestion
This is where *Deferrable Operators* can be used. When it has nothing to do but wait, a deferrable operator can suspend itself and free up the worker for other processes.  When an operator defers, execution moves to the triggerer, where the trigger specified by the operator will run, and can do the polling or waiting required by the operator.  When when the trigger finds what it's looking for, it sends a signal for the operator to resume its execution. During the deferred phase of execution, since work has been offloaded to the triggerer, the task no longer occupies a slot on the worker,  and you will have freed up some capacity. By default, deferred tasks don't use pool slots. If you would like them to, you can change this by editing the pool in question.
```

Author: Thanks so much for the edits! I took a stab at it based on your suggestions - definitely helped me get closer.

> This is where *Deferrable Operators* can be used. A deferrable operator can suspend itself and free up the worker for other processes when certain conditions are met. A pre-defined *Trigger* then resumes the deferred operator. As a result, while it is suspended (deferred) and isn't using worker slot, your cluster has fewer resources wasted on idle operators or sensors. By default, deferred tasks don't use pool slots. If you would like them to, you can change this by editing the pool in question.

Author: Awesome - your suggestion makes sense to me - there were a couple typos that I couldn't address in your comment, so had to make a follow up edit.
```suggestion
This is where *Deferrable Operators* can be used. When it has nothing to do but wait, a deferrable operator can suspend itself and free up the worker for other processes. When an operator defers, execution moves to the triggerer, where the trigger specified by the operator will run, and the trigger can do the polling or waiting required by the operator. Then, when the trigger finishes polling or waiting, it sends a signal for the operator to resume its execution. During the deferred phase of execution, since work has been offloaded to the triggerer, the task no longer occupies a worker slot, and you have more free workload capacity. By default, deferred tasks don't use pool slots. If you would like them to, you can change this by editing the pool in question.
```

> This is where *Deferrable Operators* can be used. When it has nothing to do but wait, a deferrable operator can suspend itself and free up the worker for other processes. When an operator defers, execution moves to the triggerer, where the trigger specified by the operator will run, and the trigger can do the polling or waiting required by the operator. Then, when the trigger finishes polling or waiting, it sends a signal for the operator to resume its execution. During the deferred phase of execution, since work has been offloaded to the triggerer, the task no longer occupies a worker slot, and you have more free workload capacity. By default, deferred tasks don't use pool slots. If you would like them to, you can change this by editing the pool in question.

Reviewer: ```suggestion
This is where *Deferrable Operators* can be used. When it has nothing to do but wait, an operator can suspend itself and free up the worker for other processes by *deferring*. When an operator defers, execution moves to the triggerer, where the trigger specified by the operator will run.  The trigger can do the polling or waiting required by the operator. Then, when the trigger finishes polling or waiting, it sends a signal for the operator to resume its execution. During the deferred phase of execution, since work has been offloaded to the triggerer, the task no longer occupies a worker slot, and you have more free workload capacity. By default, tasks in a deferred state don't occupy pool slots. If you would like them to, you can change this by editing the pool in question.
```

main point of this suggestion is essentially that... _any_ operator can _defer_.... what _defines_ a deferrable operator is the mere fact _that_ it defers.  So, the change from "a deferrable operator can defer" to "an operator can defer" seems a little more consistent.

Reviewer: also "deferred tasks" could make it sound like the task itself is ether "deferred" or "not deferred", which isn't what we mean. Changing to "tasks in a deferred state" makes it clear that not occupying a slot is a temporary phenomenon, only while task is in deferred state.

Author: Thank you so much - these fixes and clarifications look great.

> *Triggers* are small, asynchronous pieces of Python code designed to run in a single Python process. Because they are asynchronous, they can all co-exist efficiently in a *triggerer*, an Airflow service similar to a scheduler or worker.

Reviewer 2: nit: Not sure I like "similar" here, that section. Maybe "in the triggerer Airflow component".

> *Triggers* are small, asynchronous pieces of Python code designed to run in a single Python process. Because they are asynchronous, they can all co-exist efficiently in a *triggerer*, an Airflow service similar to a scheduler or worker.

Author: ```suggestion
*Triggers* are small, asynchronous pieces of Python code designed to run in a single Python process. Because they are asynchronous, they can all co-exist efficiently in the *triggerer* Airflow component.
```

* The scheduler queues the task to resume on a worker node.

You can either use pre-written deferrable operators as a DAG author or write your own. Writing them, however, requires that they meet certain design criteria.

Using Deferrable Operators
--------------------------

If you want to use pre-written deferrable operators that come with Airflow, such as ``TimeSensorAsync``, then you only need to complete two steps:

* Ensure your Airflow installation runs at least one ``triggerer`` process, as well as the normal ``scheduler``
* Use deferrable operators/sensors in your DAGs

Airflow automatically handles and implements the deferral processes for you.

If you're upgrading existing DAGs to use deferrable operators, Airflow contains API-compatible sensor variants, like ``TimeSensorAsync`` for ``TimeSensor``. Add these variants into your DAG to use deferrable operators with no other changes required.

Note that you can't use the deferral ability from inside custom PythonOperator or TaskFlow Python functions. Deferral is only available to traditional, class-based operators.

.. _deferring/writing:

Writing Deferrable Operators
----------------------------

> Writing a deferrable operator requires more configuration than updating your DAGs with pre-written operators. There are some main points to consider:

Reviewer 2: ```suggestion
When writing a deferrable operators these are the main points to consider:
```

If you are worried about carryover context from the last section, let's do this instead. Writing operators will naturally be more complex than using pre-written ones - I see no value in being more verbose here.

* Your operator must defer itself with a trigger. You can use a trigger included in core Airflow, or you can write a custom one.

> * Your operator will be stopped and removed from its worker while deferred, and no state persists automatically. You can persist state by instructing Airflow to resume the operation at a certain method or by passing certain kwargs.

Reviewer 2: ```suggestion
* Your operator will be stopped and removed from its worker while deferred, and no state persists automatically. You can persist state by instructing Airflow to resume the operator at a certain method or by passing certain kwargs.
```

Triggering Deferral
~~~~~~~~~~~~~~~~~~~

If you want to trigger deferral, at any place in your operator, you can call ``self.defer(trigger, method_name, kwargs, timeout)``. This raises a special exception for Airflow. The arguments are:

> * ``trigger``: An instance of a trigger that you want to defer. It will be serialized into the database.

Reviewer 2: ```suggestion
* ``trigger``: An instance of a trigger that you want to defer to. It will be serialized into the database.
```

We are deferring the operator, not the trigger. The trigger is the responsible party now. So maybe this?

> * ``method_name``: The method name of your operator that you want Airflow to call when it resumes.

Reviewer 2: ```suggestion
* ``method_name``: The method name on your operator that you want Airflow to call when it resumes.
```

"on" is correct here. Or "in". It's a reference to the method in the task object.

Triggers can be as complex or as simple as you want, provided they meet the design constraints. They can run in a highly-available fashion, and are auto-distributed among hosts running the triggerer. We encourage you to avoid any kind of persistent state in a trigger. Triggers should get everything they need from their ``__init__``, so they can be serialized and moved around freely.

If you are new to writing asynchronous Python, be very careful when writing your ``run()`` method. Python's async model means that code can block the entire process if it does not correctly ``await`` when it does a blocking operation. Airflow attempts to detect process blocking code and warn you in the triggerer logs when it happens. You can enable extra checks by Python by setting the variable ``PYTHONASYNCIODEBUG=1`` when you are writing your trigger to make sure you're writing non-blocking code. Be especially careful when doing filesystem calls, because if the underlying filesystem is network-backed, it can be blocking.

High Availability
-----------------

> Triggers are designed to work in high availability (HA) architecture. If you want to run a high availability setup, run multiple copies of ``triggerer`` on multiple hosts. Much like ``scheduler``, they automatically co-exist with correct locking and HA.

Reviewer 2: ```suggestion
Triggers are designed to work in a high availability (HA) architecture. If you want to run a high availability setup, run multiple copies of ``triggerer`` on multiple hosts. Much like ``scheduler``, they automatically co-exist with correct locking and HA.
```

nit
